docs/Debugging.md

```diff
@@ -0,0 +1,81 @@
+
+The instructions below assume that Open Enclave was installed in the default directory `/opt/openenclave`.
+If you used a different directory, run `. <OE_PREFIX>/share/openenclave/openenclaverc` first.
+
+### Build mode
+
+Debugging enclave code (SGX-LKL as well as the application code) is only possible in debug and non-release builds of SGX-LKL.
+
+### Encrypted disks
+
+TODO add note on how to handle situations when disk keys are retrieved from a remote service that would not release them to debug/non-release builds of SGX-LKL (either re-create disks unencrypted, or, if available locally, hard-code key in app config for local debugging)
+
+### Confidential Containers
+
+The instructions below assume that disk images and configuration are available outside of a Confidential Container. Debugging of Confidential Containers (via Docker) is not supported currently.
+
+## Option A: Debugging with an SGX-LKL installation
+
+The instructions below assume that SGX-LKL was installed in `/opt/sgx-lkl`.
+
+To debug an application, invoke `sgx-lkl-gdb` as follows:
+```
+/opt/sgx-lkl/bin/sgx-lkl-gdb --args /opt/sgx-lkl/bin/sgx-lkl-run-oe --hw-debug root_disk.img /app
+```
+See `sgx-lkl-run-oe --help` for how to run using host and app config files.
+
+## Option B: Debugging with an SGX-LKL source build
+
+This assumes you want to debug SGX-LKL directly from the source tree without installing it.
+
+To debug an application, invoke `sgx-lkl-gdb` as follows:
+```
+tools/gdb/sgx-lkl-gdb --args build/sgx-lkl-run-oe --hw-debug root_disk.img /app
+```
+See `sgx-lkl-run-oe --help` for how to run using host and app config files.
+
+## Debugging advice
+
+### Visual Studio Code
+
+Visual Studio Code allows to use a custom `gdb` command for debugging.
+To use `sgx-lkl-gdb`, add `"miDebuggerPath": "/opt/sgx-lkl/bin/sgx-lkl-gdb"` (change path if necessary) in your `.vscode/launch.json` file.
+Tip: Use the "Visual Studio Code Remote - SSH" extension to debug on a remote VM with SGX support.
+
+### Run mode
+
+Apart from obvious differences (security, performance, ...) there are subtle differences between SGX-LKL software mode (`--sw-debug`) and hardware mode (`--hw-debug`).
+If possible, debugging should be done in hardware mode to be as close to a production environment as possible.
+
+### Ignore SIGILL
+
+In some cases `sgx-lkl-gdb` still incorrectly stops at illegal instructions that are emulated inside the enclave (see e.g. https://github.com/lsds/sgx-lkl/issues/167). To avoid that, run `handle SIGILL pass nostop noprint` as first command in the active `gdb` session.
+
+### Optimized code
+
+In general, debugging optimized code is not recommended.
+For an optimal debugging experience, SGX-LKL should be built in debug mode.
+If the application code should be debugged as well and it is native code,
+then the application should be built without optimizations and with debug symbols.
```

Comment:
We should mention somewhere at the top that debugging non-native applications is currently not possible. Debugging non-native language runtimes is also challenging because they typically are not built with symbols, e.g., https://github.com/lsds/sgx-lkl/issues/245

Reply:
I added a subsection at the top about non-native applications. I don't think we have to go into language runtimes as that's a very advanced topic and people would probably know what they are doing in that case.